### PR TITLE
Move environment-mac.yaml to Python 3.9 and patch dream.py for Macs.

### PR DESCRIPTION
**Pull request by corajr**
I'm using stable-diffusion on a 2022 Macbook M2 Air with 24 GB unified memory.
I see this taking about 2.0s/it.

I've moved many deps from pip to conda-forge, to take advantage of the
precompiled binaries. Some notes for Mac users, since I've seen a lot of
confusion about this:

One doesn't need the `apple` channel to run this on a Mac-- that's only
used by `tensorflow-deps`, required for running tensorflow-metal. For
that, I have an example environment.yml here:

https://developer.apple.com/forums/thread/711792?answerId=723276022#723276022

However, the `CONDA_SUBDIR=osx-arm64` environment variable *is* needed to
ensure that you do not run any Intel-specific packages such as `mkl`,
which will fail with [cryptic errors](https://github.com/CompVis/stable-diffusion/issues/25#issuecomment-1226702274)
on the ARM architecture and cause the environment to break.

I've also added a comment in the env file about 3.10 not working yet.
When it becomes possible to update, those commands run on an osx-arm64
machine should work to determine the new version set.

Here's what a successful run of dream.py should look like:

```
$ python scripts/dream.py --full_precision                                                                                                           SIGABRT(6) ↵  08:42:59
* Initializing, be patient...

Loading model from models/ldm/stable-diffusion-v1/model.ckpt
LatentDiffusion: Running in eps-prediction mode
DiffusionWrapper has 859.52 M params.
making attention of type 'vanilla' with 512 in_channels
Working with z of shape (1, 4, 32, 32) = 4096 dimensions.
making attention of type 'vanilla' with 512 in_channels
Using slower but more accurate full-precision math (--full_precision)
>> Setting Sampler to k_lms
model loaded in 6.12s

* Initialization done! Awaiting your command (-h for help, 'q' to quit)
dream> "an astronaut riding a horse"
Generating:   0%|                                                                                                                                                                         | 0/1 [00:00<?, ?it/s]/Users/corajr/Documents/lstein/ldm/modules/embedding_manager.py:152: UserWarning: The operator 'aten::nonzero' is not currently supported on the MPS backend and will fall back to run on the CPU. This may have performance implications. (Triggered internally at /Users/runner/work/_temp/anaconda/conda-bld/pytorch_1662016319283/work/aten/src/ATen/mps/MPSFallback.mm:11.)
  placeholder_idx = torch.where(
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 50/50 [01:37<00:00,  1.95s/it]
Generating: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1/1 [01:38<00:00, 98.55s/it]
Usage stats:
   1 image(s) generated in 98.60s
   Max VRAM used for this generation: 0.00G
Outputs:
outputs/img-samples/000001.1525943180.png: "an astronaut riding a horse" -s50 -W512 -H512 -C7.5 -Ak_lms -F -S1525943180
```

### CODE REVIEW

**Review by lstein — APPROVED** [review]
LGTM.

**Comment by jroxendal** [comment]
This didn't work for me, here's my output:

```
dream> clouds
Generating:   0%|                                                              | 0/1 [00:00<?, ?it/s]/Users/johan/dev/stable-diffusion/ldm/modules/embedding_manager.py:152: UserWarning: The operator 'aten::nonzero' is not currently supported on the MPS backend and will fall back to run on the CPU. This may have performance implications. (Triggered internally at /Users/runner/work/_temp/anaconda/conda-bld/pytorch_1662016319283/work/aten/src/ATen/mps/MPSFallback.mm:11.)
  placeholder_idx = torch.where(
                                                                                                    loc("mps_add"("(mpsFileLoc): /AppleInternal/Library/BuildRoots/20d6c351-ee94-11ec-bcaf-7247572f23b4/Library/Caches/com.apple.xbs/Sources/MetalPerformanceShadersGraph/mpsgraph/MetalPerformanceShadersGraph/Core/Files/MPSGraphUtilities.mm":219:0)): error: input types 'tensor<2x1280xf32>' and 'tensor<*xf16>' are not broadcast compatible
LLVM ERROR: Failed to infer result type(s).
Abort trap: 6
/Users/johan/opt/anaconda3/envs/ldm/lib/python3.9/multiprocessing/resource_tracker.py:216: UserWarning: resource_tracker: There appear to be 1 leaked semaphore objects to clean up at shutdown
  warnings.warn('resource_tracker: There appear to be %d '
```

**Comment by heurihermilab** [comment]
Not working for me either.  Using txt2img.py prompts that worked before today (with v 3.10), either hangs or runs many times more slowly, enough that the progress bar never updates after usual startup pause.  Memory use through the roof, freezing UI in other apps, had to kill it just to type this.  Have updated to pytorch-nightly with no change.

Following is response after initialization and before progress bar:

```
[...]/stable-diffusion/ldm/modules/embedding_manager.py:152: UserWarning: The operator 'aten::nonzero' is not currently supported on the MPS backend and will fall back to run on the CPU. This may have performance implications. (Triggered internally at /Users/runner/work/_temp/anaconda/conda-bld/pytorch_1662016319283/work/aten/src/ATen/mps/MPSFallback.mm:11.)
  placeholder_idx = torch.where(
```

Have confirmed the Python 3.9 is arm64.

Environment: MBP 14", M1 Pro, 16GB, latest OS, running miniforge with a base of Python 3.10.6.  Ran conda clean and reset while debugging.

**Comment by corajr** [comment]
@jroxendal you need to specify the `--full_precision` flag, mixed-precision doesn't work on Macs.

@heurihermilab concerning. I think this is due to pytorch-nightly -- I thought it should be faster because `aten::index.Tensor` was not implemented for MPS until recently (https://github.com/pytorch/pytorch/issues/77959). However it looks like the performance is actually worse.

In the [previous environment-mac.yaml](https://github.com/lstein/stable-diffusion/blob/2aa83932728242b05248fada8447506e2fbed9f0/environment-mac.yaml) with Python 3.10, I see a peak memory usage of 11.50 GB and iteration times of around 1.78s/it for 1 sample. With Python 3.9 and pytorch-nightly as of this PR, memory peak is 13.70 GB and iteration time 2.0s/it (a clear regression). When switching the channel to stable `pytorch` in this Python 3.9 env, I get 11.30 GB and 1.78s/it, just as with Python 3.10.

I see there is another PR https://github.com/lstein/stable-diffusion/pull/268 about to update this again. I will try it out and see whether some of the conda-forge binary goodness can be preserved.

**Comment by heurihermilab** [comment]
Supplemental: 

After a thorough rebuilding with the updated README-Mac-MPS.md instructions, I have been able to get both dream.py and txt2img.py to run after adding pytorch-nightly (not entirely certain this was required).  Alas, all I get is black.

I've also seen some weirdness where it seems like TWO instances of 3.9 get spawned, which then compete to grab all of memory, which may be what's behind the slowdown I reported.

I will update as events merit...

**Comment by Birch-san** [comment]
for what it's worth: I use Python 3.10 for stable-diffusion just fine. I haven't dried the lstein fork and dream.py, but txt2img.py and img2img.py work fine on 3.10 on M1, in my experience.

**Comment by corajr** [comment]
@Birch-san Python 3.10 works for me too, but the conda environment has to compile a ton of packages because so many are listed under pip -- that's the main reason I want to move these deps to use the conda-forge precompiled versions.

I realized why I was unable to solve the Python 3.10 environment. The recipe of the conda-forge package for `tokenizers` has never been updated to build for Python 3.10, so `transformers` can't find a suitable version. I just created https://github.com/conda-forge/tokenizers-feedstock/pull/51 to rectify this, and once that's merged the environment should be solvable again with `transformers` included.

In the meantime I suspect the `environment-mac.yaml` on main of this repo is probably broken with the merge of #268, I'll try to get it into a working state again.
